STIPS/STIP-004.md

```diff
@@ -0,0 +1,642 @@
+* Multiple positive external positions are the same as a single positive external position because we are summing up all the positive external positions in `DebtIssuanceModule#_getTotalIssuanceUnits`.
+
+
+#### Note (for redeeming): 
+
+* SetToken's are burned at the very beginning of the DebtIssuanceModule#redeem() function, hence in columns 5,6,7,8 (count starting with 1) of the above table, setToken.totalSupply() is `s - r'`, where `r'` is the redeem quantity **without** fees.
+* After column 8, fees are minted to the fee receiver and setToken’s total supply becomes `s - r`.
+
+## Feasibility Analysis
+
+### Option 1:
+
+Instead of over-optimizing our checks to check for balances which effectively prevent both undercollaterlization and overcollaterlization. We can just have checks to prevent undercollaterlization. The change is contained within the smart contracts and doesn't bubble up the stack leading to bigger changes in our backend.
+
+The proposed checks for undercollateralization are performed right after the transfer happens, so that we revert immediately and save gas. When token is transferred in, we wish to check the right amount of equity/debt was passed in before we try to transfer it out to external protocols. When token is transferred out, we validate that we are not undercollateralized after the transfer.
+
+### Proposed checks during Issuance
+* In `_resolveEquityPositions` we can validate balances after the equity is transferred but before the external position hooks are called, because
+* Proposed check:
+
+```javascript
+// pseudocode
+uint256 defaultPositionUnit = setToken.getDefaultPositionRealUnit(component)
+uint256 cumulativeEquity = defaultPositionUnit
+address[] externalPositionModules = setToken.getExternalPositionModules(component)
+for externalPositionModule in externalPositionModules {
+   uint256 externalPositionUnit = setToken.getExternalPositionRealUnit(component, externalPositionModule)
+   if (externalPositionUnit > 0)
+      cumulativeEquity = cumulativeEquity.add(externalPositionUnit);
+}
+uint256 newBalance = component.balanceOf(setToken)
+require(newBalance > s * defaultPositionUnit + i * cumulativeEquity)
+```
+
+* In `_resolveDebtPositions` we validate set collateralization after debt is transferred back to the user.
+* Proposed check:
+
+```javascript
+// pseudocode
+defaultPositionUnit = setToken.getDefaultPositionRealUnit(component)
+newBalance = component.balanceOf(setToken)
+require(newBalance >= (s+i) * defaultPositionUnit)
+```
+
+
+### Proposed checks during Redeeming
+
+* In `_resolveDebtPositions` we can check balances after debt is transferred but before the external position hooks are called.
+* Check:
+
+```javascript
+// pseudocode
+defaultPositionUnit = setToken.getDefaultPositionRealUnit(component)
+cumulativeDebt = 0
+externalPositionModules = setToken.getExternalPositionModules(component)
+For each externalPositionModule in externalPositionModules
```

Comment:
Here we may be able to imply `cumulativeDebt` by passing in `totalDebtAmount` and dividing by `r`

Reply:
We need not divide by `r`. 
In the actual implementation, we pass in the pre-calculated `cumulativeDebt` value via the `compnentQuantity` parameter.